lib/matplotlib/collections.py

```diff
@@ -2047,6 +2048,43 @@ def set_paths(self):
         self._paths = self._convert_mesh_to_paths(self._coordinates)
         self.stale = True
 
+    def set_array(self, A):
+        """
+        Validates the dimensions of the array to be set based on 'shading'.
+        """
+        height, width = self._coordinates.shape[0:-1]
+        misshapen_data = False
+        faulty_data = False
+
+        if self._shading == 'flat':
+            h, w = height-1, width-1
+        else:
+            h, w = height, width
+
+        if A is not None:
+            if A.ndim == 1:
+                if A.size != (h*w):
+                    faulty_data = True
+            elif A.shape != (h, w):
+                if A.size == (h * w):
```

Comment:
When writing the docstring, I realized that we should (as before) accept non-array input that is array-like (i.e. can be converted to an array). The dimension/shape checks currently fail on that.

One way is :

```suggestion
            shape = np.shape(A)
            if len(shape) == 1:
                if shape[0] != h * w:
                    faulty_data = True
            elif shape != (h, w):
                if np.prod(shape) == h * w:
```

The alternative would be to do the array conversion early, but that would mean more or less applying `safe_masked_invalid()` again as in the super method, and that would result in duplicate copying.

After all, after the deprecation period, the checks simplify to
```
if shape != (h, w) and shape != (h*w,):
     faulty_data ...
```

Reply:
The first way was how it was written originally, I'll revert to that instead